sql/core/src/main/scala/org/apache/spark/sql/MergeIntoWriter.scala

```diff
@@ -0,0 +1,316 @@
+   *   - `updateAll`: Update all the target table fields with source dataset fields.
+   *   - `update(Map)`: Update all the target table records while changing only
+   *     a subset of fields based on the provided assignment.
+   *   - `delete`: Delete all the target table records.
+   *
+   * @return a new `WhenNotMatchedBySource` object.
+   */
+  def whenNotMatchedBySource(): WhenNotMatchedBySource[T] = {
+    new WhenNotMatchedBySource[T](this, None)
+  }
+
+  /**
+   * Initialize a `WhenNotMatchedBySource` action with a condition.
+   *
+   * This `WhenNotMatchedBySource` action will be executed if and only if the specified `condition`
+   * is satisfied.
+   *
+   * This `WhenNotMatchedBySource` can be followed by one of the following merge actions:
+   *   - `updateAll`: Update all the target table fields with source dataset fields.
+   *   - `update(Map)`: Update all the target table records while changing only
+   *     a subset of fields based on the provided assignment.
+   *   - `delete`: Delete all the target table records.
+   *
+   * @param condition a `Column` representing the condition to be evaluated for the action.
+   * @return a new `WhenNotMatchedBySource` object configured with the specified condition.
+   */
+  def whenNotMatchedBySource(condition: Column): WhenNotMatchedBySource[T] = {
+    new WhenNotMatchedBySource[T](this, Some(condition.expr))
+  }
+
+  /**
+   * Executes the merge operation.
+   */
+  def merge(): Unit = {
+    if (matchedActions.isEmpty && notMatchedActions.isEmpty && notMatchedBySourceActions.isEmpty) {
+      throw QueryExecutionErrors.mergeIntoAPIError()
+    }
+
+    val merge = MergeIntoTable(
+      UnresolvedRelation(tableName),
+      logicalPlan,
+      on.get.expr,
+      matchedActions,
+      notMatchedActions,
+      notMatchedBySourceActions)
+    val qe = sparkSession.sessionState.executePlan(merge)
+    qe.assertCommandExecuted()
+  }
+}
+
+/**
+ * A class for defining actions to be taken when matching rows in a DataFrame during
+ * an update operation.
+ *
+ * @param mergeIntoWriter   The MergeIntoWriter instance responsible for writing data to a
+ *                          target DataFrame.
+ * @param condition         An optional condition Expression that specifies when the actions
+ *                          should be applied.
+ *                          If the condition is None, the actions will be applied to all matched
+ *                          rows.
+ *
+ * @tparam T                The type of data in the MergeIntoWriter.
+ */
+case class WhenMatched[T] (mergeIntoWriter: MergeIntoWriter[T], condition: Option[Expression]) {
+  /**
+   * Specifies an action to update all matched rows in the DataFrame.
+   *
+   * @return The MergeIntoWriter instance with the update all action configured.
+   */
+  def updateAll(): MergeIntoWriter[T] = {
+    mergeIntoWriter.matchedActions = mergeIntoWriter.matchedActions :+ UpdateStarAction(condition)
```

Comment:
do we need to consider thread safety here?

Reply:
Each invocation of `mergeInto` would create a new instance of `MergeIntoWriter`.  I think we don't need to consider thread safe here.

Reply:
BTW, I think we can make the code simpler by adding a new method in `MergeIntoWriter`:
```
def withNewMatchedAction(action): MergeIntoWriter = {
  this.matchedActions = this.matchedActions :+ ...
  this
}
```

and here we just need to do
```
mergeIntoWriter.withNewMatchedAction(...)
```

Reply:
I didn't see this last night. Fixed just now.